consensus/src/lib.rs

```diff
@@ -1,27 +1,624 @@
+        let merkle_tree_root = block.get_merkle_root();
+        match calculate_tx_merkle_root(block.get_transactions()) {
+            Ok(merkle_tree) => {
+                if merkle_tree_root != merkle_tree {
+                    return Err(BlockError::Unknown);
+                }
+            }
+            Err(_merkle_error) => {
+                // TODO: Should we return additional error information?
+                return Err(BlockError::Unknown);
+            }
+        }
+
+        // Witness merkle root
+        let witness_merkle_root = block.get_witness_merkle_root();
+        match calculate_witness_merkle_root(block.get_transactions()) {
+            Ok(witness_merkle) => {
+                if witness_merkle_root != witness_merkle {
+                    return Err(BlockError::Unknown);
+                }
+            }
+            Err(_merkle_error) => {
+                // TODO: Should we return additional error information?
+                return Err(BlockError::Unknown);
+            }
+        }
+
+        match &block.get_prev_block_id() {
+            Some(block_id) => {
+                let previous_block = self
+                    .db_tx
+                    .get_block_index(&Id::<Block>::from(block_id))?
+                    .ok_or(BlockError::NotFound)?;
+                // Time
+                let block_time = block.get_block_time();
+                if previous_block.get_block_time() > block_time {
+                    return Err(BlockError::Unknown);
+                }
+                if i64::from(block_time) > time::get() {
+                    return Err(BlockError::Unknown);
+                }
+            }
+            None => (),
+        }
+
+        self.check_transactions(block)?;
+        Ok(())
+    }
+
+    #[allow(dead_code)]
+    fn check_consensus(&self, block: &Block) -> Result<(), BlockError> {
+        let _consensus_data = block.get_consensus_data();
+        // TODO: PoW is not in master at the moment =(
+        Ok(())
+    }
+
+    #[allow(dead_code)]
+    fn check_transactions(&self, _block: &Block) -> Result<(), BlockError> {
+        //TODO: Must check for duplicate inputs (see CVE-2018-17144)
+        //TODO: Size limits
+        //TODO: Check signatures
+        Ok(())
+    }
+
+    #[allow(dead_code)]
+    fn check_block(&self, block: &Block) -> Result<(), BlockError> {
+        self.check_consensus(block)?;
+        self.check_block_detail(block)?;
+        Ok(())
+    }
+
+    fn is_block_in_main_chain(&self, block_index: &BlockIndex) -> bool {
+        block_index.next_block_id.is_some()
+            || match self.db_tx.get_best_block_id().ok().flatten() {
+                Some(ref block_id) => block_index.get_block_id() == block_id,
+                None => false,
+            }
+    }
+
+    /// Mark new block as an orphan
+    /// Ok(()) - Added
+    /// Err(BlockError) - StorageFailure  
+    fn new_orphan_block(&mut self, block: Block) -> Result<(), BlockError> {
```

Comment:
Should this take reference instead and copy/clone the block only when it's actually necessary?

Reply:
@altonen No, if block is orphan, we should return error and don't accept this block. Should be unable to use block after the `new_orphan_block`.

Let's continue in a new PR: https://github.com/mintlayer/mintlayer-core/pull/124